Implemented `List.init`

Richard-Degenne opened the pull request:
# Summary

I was amazed to see that `List.init` didn't exist in the standard library, and thought that it should be fixed.

This PR implements `List.init`.

# Description

`List.init` works exactly like `Array.init` does. It allows the user to create a list given its size and a function which will be mapped to the new list.

The size must be positive. An `Invalid_argument` exception is raised if the size is negative. Giving 0 as a size will result in the new list being empty.

## Signature

    val init : int -> (int -> 'a) -> 'a list

## Examples

### Correct usage

    List.init 10 (fun x -> x);;
    - : int list = [0; 1; 2; 3; 4; 5; 6; 7; 8; 9]

    List.init 5 (fun x -> 2*(x+1));;
    - : int list = [2; 4; 6; 8; 10]

    List.init 5 (string_of_int);;
    - : bytes list = ["0"; "1"; "2"; "3"; "4"]

### Incorrect usage

    List.init (-1) (fun x -> x);;
    Exception: (Invalid_argument List.init).

# Design

As you can see [here](https://github.com/ocaml/ocaml/compare/trunk...Richard-Degenne:list-init?expand=1#diff-59bc62b6e16ab262d703d043e2d1d51b), `List.init` is tail-recursive; it uses a auxiliary function to apply the given function to every element before appending them to an accumulator, which is then reversed to return the new list in the correct order.

# Documentation

As you cas see [here](https://github.com/ocaml/ocaml/compare/trunk...Richard-Degenne:list-init?expand=1#diff-d13fc29ec4bc8a3ef7dd076bc6be16f9), documentation has been added to the function in the `.mli` file.

# Tests

Unit tests have been added to a new folder `testsuite/tests/list-functions` because I was unable to find any tests for other `List` functions.

    $ make one DIR=tests/list-functions                                                                                                          

    Running tests from 'tests/list-functions' ...
     ... testing 'test.ml': ocamlc ocamlopt => passed

Commit:
Implemented `List.init`

Commit:
Wrote unit tests

dra27 commented:
Thanks for this - especially for including tests! A few things:

1. Testsuite files do not need copyright headers (some have them).
2. I'm surprised that we don't have any List tests already (unless I'm being blind) - but could you instead name the directory lib-list for consistency with other stdlib test directories?
3. I'm personally not convinced by the choice of name. `init` has a totally different in Haskell and it also, to me, sounds very imperative for a functional structure - especially as it implies mutability. In Standard ML this function is called `tabulate`. However, it is the name chosen by both Batteries and Core.

dra27 commented:
You will also need a `Changes` file entry and there are some testsuite issues to address. Looking only at the list of fails, it looks like you are missing the equivalent change in `ListLabels` - I expect that the translprim test is a renumbering problem - see what the difference between the result and reference files is.

Richard-Degenne commented:
> Testsuite files do not need copyright headers (some have them).
> I'm surprised that we don't have any List tests already (unless I'm being blind) - but could you instead name the directory lib-list for consistency with other stdlib test directories?

I'll correct that right away, while keeping an eye out for those missing tests.

> I'm personally not convinced by the choice of name. `init` has a totally different in Haskell and it also, to me, sounds very imperative for a functional structure - especially as it implies mutability. In Standard ML this function is called `tabulate`. However, it is the name chosen by both Batteries and Core.

`init` is also the name chosen in the `Array` module. It may be better to stay with that name, even though, as you stated, it seems to imply mutability.

> You will also need a Changes file entry.

Should I add the entry in the *Next version* or in the *Next minor version*?

> And there are some testsuite issues to address.

I'll look into that too.

Anyway, thanks a lot for your feedback. This is my first time contributing and having a bit of help is very helpful.

braibant commented on stdlib/list.mli:
```diff
@@ -73,6 +73,16 @@ val nth_opt: 'a list -> int -> 'a option
+   In other terms, [List.init n f] tabulates the results of [f]
+   applied to the integers [0] to [n-1].
+
+   Raise [Invalid_argument] if [n < 0] or [n > Sys.max_array_length].
```
This seems wrong. The code does not check that `n < Sys.max_array_length`

Richard-Degenne replied:
Agreed.

braibant commented on stdlib/list.ml:
```diff
@@ -56,6 +56,14 @@ let rec rev_append l1 l2 =
+let init n f =
+  if n = 0 then [] else
+  if n < 0 then invalid_arg "List.init" else
+  let rec init_aux acc i =
```
Isn't this going to allocate the closure for `init_aux` for each call to init (without flamba)? What about:

```
let init_aux acc i f = 
....
;;

let init n f =
   ... (* checks *);
   rev (init_aux [] 0 f)
;;
```

Richard-Degenne replied:
Agreed.

mshinwell commented:
I haven't looked at the code, but would like to support the inclusion of this function.  The next version of Flambda has a lot of `Array.to_list (Array.init ...)` at the moment.

Commit:
Fixed failing `lib-stdlabels` unit tests

Commit:
Included GitHub review comments

See [here](https://github.com/ocaml/ocaml/pull/1034/files/db41747f942f77d4ac14a2551a6489095eacb062#r100268823)
and [here](https://github.com/ocaml/ocaml/pull/1034/files/db41747f942f77d4ac14a2551a6489095eacb062#r100268386)
for details.

Commit:
Renamed unit tests + Changes entry

dra27 commented:
It would be "Next version" - if this can be polished before Wednesday, it could be ready for 4.05.

@alainfrisch - any objections, once fixes made?

Octachron commented on stdlib/list.mli:
```diff
@@ -73,6 +73,16 @@ val nth_opt: 'a list -> int -> 'a option
+
+   Raise [Invalid_argument] if [n < 0] or [n > Sys.max_array_length].
+   If the return type of [f] is [float], then the maximum
+   size is only [Sys.max_array_length / 2].*)
```
This paragraph is copy-pasted from `Array.init` and should not be here.

dra27 replied:
It does need part of it - the `[n < 0]` bit is true.

Octachron commented on stdlib/list.ml:
```diff
@@ -56,6 +56,14 @@ let rec rev_append l1 l2 =
+  let rec init_aux acc i =
+    if i >= n then acc else
+    init_aux ((f i)::acc) (i+1)
+  in rev (init_aux [] 0)
```
Do we really need the property that `f` is called in the order `0` to `n-1`?
If the answer is no, there is no reason to reverse the list at the end rather than construct the list in reverse order, i.e 
```OCaml
let init_aux f l n = if n = 0 then f 0 :: l else init_aux f (f n :: l) (n-1)
```
If the answer is yes, this property ought to be documented.

Richard-Degenne replied:
Agreed.

dra27 replied:
I wondered whether really we should be adding scan_left and scan_right at this point, but that's a bit too SML-y for the OCaml standard library! It seems better to me to maintain the duality with `Array.init` - so document the fact.

That said, `Array.init` doesn't note it either...

Commit:
Included GitHub review comment

See [here](https://github.com/ocaml/ocaml/pull/1034#discussion_r100311362) for details.

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -73,6 +73,14 @@ val nth_opt: 'a list -> int -> 'a option
 val rev : 'a list -> 'a list
 (** List reversal. *)
 
+val init : int -> (int -> 'a) -> 'a list
```
For consistency with other functions, I think the signature should be

```ocaml
val init : len:int -> f:(int -> 'a) -> 'a list
```

Richard-Degenne replied:
Agreed.

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -73,6 +73,14 @@ val nth_opt: 'a list -> int -> 'a option
 (** List reversal. *)
 
+val init : int -> (int -> 'a) -> 'a list
+(** [List.init n f] returns a fresh list of length [n],
```
Replace with comment above.

dra27 commented on stdlib/list.mli:
```diff
@@ -73,6 +73,14 @@ val nth_opt: 'a list -> int -> 'a option
+   In other terms, [List.init n f] tabulates the results of [f]
+   applied to the integers [0] to [n-1].
+
+   Raise [Invalid_argument] if [n < 0].*)
```
Sorry for more comments on this comment. I think the copy-paste from Array has lost sight of how this function should be documented, especially now that it builds the list in reverse.

```ocaml
(** [List.init n f] is [f 0; f 1; ... ; f (n - 1)].

    @raise Invalid_argument if [n < 0].

    @since 4.05.0
*)
```

I can't decide whether we should be explicit that the list is constructed in reverse - I *think* that the order of evaluation of `[f 0; f 1]` is itself not specified in OCaml, which would imply that not mentioning it is OK?

Richard-Degenne replied:
A workaround would be to define both `init` and `init_rev`. But is it really interesting?

Richard-Degenne replied:
`Core`'s `List.init` builds the list in reverse too:

```
open Core.Std;;
List.init 10 ~f:(fun x -> print_endline (string_of_int x));;
9
8
7
6
5
4
3
2
1
0
- : unit list = [(); (); (); (); (); (); (); (); (); ()]
```

dra27 reviewed (commented):
More changes to the comments, please!

Richard-Degenne commented:
I can't find what's wrong with `testsuite/tests/translprim/comparison.ml`. The test file itself is pretty confusing and the test output doesn't make much sense to me.

```
$ diff testsuite/tests/translprim/comparison_table.ml.{reference,result}

246c246
<                 (apply (field 15 (global List!)) (apply uncurry f)
---
>                 (apply (field 16 (global List!)) (apply uncurry f)
294c294
<                     (apply (field 15 (global List!))
---
>                     (apply (field 16 (global List!))
```

Any idea?

dra27 commented:
It looks fine - just replace the .reference file with the .result file. You've added a member to the List structure which changes the ordinal numbers for the functions when looking further down the compiler pipeline.

Richard-Degenne commented:
Ok, that's what I thought. Thanks.

Commit:
Modified `ListLabels.init` signature

`List.init` has been update to reflect the changes.

Commit:
Uptated existing unit tests to match implementation

Commit:
Updated documentation

dra27 commented on stdlib/list.mli:
```diff
@@ -73,6 +73,11 @@ val nth_opt: 'a list -> int -> 'a option
+val init : int -> (int -> 'a) -> 'a list
+(** [List.init len f] is [f 0; f 1; ...; f (n-1)].
+
+   Raise [Invalid_argument] if [n < 0].*)
```
It is better to be using `@raise Invalid_argument ...` and please do include the `@since 4.05.0` line

Richard-Degenne replied:
Agreed.

gasche commented:
I believe that calling the function left-to-right, from `0` to `n - 1`, is an important property of a no-surprise implementation.

Richard-Degenne commented:
The answer seems to be up to us, here. As I said yesterday [here](https://github.com/ocaml/ocaml/pull/1034#discussion_r100329841), `Core` calls `f` from `n-1` to `0` and [the function's documentation](https://ocaml.janestreet.com/ocaml-core/111.17.00/doc/core/#Std.List) doesn't mention it.

gasche commented:
Well, my personal opinion is that Core made the wrong design choice here. Users will assume that evaluation is left-to-right, and they will be surprised (or not notice and have bugs in their programs; of course they should not have effects that rely on evaluation order, but in practice they do). Surprise is bad -- in this context.

Richard-Degenne commented:
I think we need more input on this.

mshinwell commented:
If anything, right-to-left order is more consistent with the rest of the compiler, which after some fixes currently in development I think will always use that order.

I don't think we should pay the penalty of twice as much allocation just for those people who don't read the documentation.  I think it is reasonable to document the order.

Richard-Degenne commented:
`Array.init` applies `f` from `0` to `n-1`, though. But an `Array` is an imperative structure, so I'm not sure whether it *has* to be consistent with `List.init`.

nojb commented:
Is it necessary to make `List.init` tail-recursive ?  Functions which are not naturally tail recursive are typically left like that in the stdlib (e.g. `List.map`, etc).

dra27 commented:
@nojb - having not initially been bothered, it does now seem strange to me for this function not to be tail-recursive (also, of course, there is `rev_map` provided - indeed, mostly the List module provides the logical alternate tail-recursive function) or at least for one not to provided (and providing two init functions feels like a lot of weight for something so simple).

@gasche - Core does it tail-recursively in reverse, but Batteries only does it left-to-right by cheating with tail-consed lists! Would you be happy if the docs stated that the order of evaluation is *unspecified*?

@mshinwell - is right-to-left for initialisation of lists specified, or am I correct that it is intentionally not specified? I know it's been the behaviour of `[f 0; f 1; f 2]` for approximately ever!

gasche commented:
I believe that most users (1) never carefully read the documentation of the function when it has the name that "obviously" gives the meaning they want, (2) will easily make assumptions about evaluation order (usually left-to-right, small to large), and (3) even if they test the behavior and see that their guess was wrong (at this point you made them lose their time and you should start feeling guilty), they will always assume that the behavior they can now observe is not going to change in the future. As a result, I am not very interested in the question of what is (un)specified, but I have a relatively firm opinion on what should be implemented.

(Note that the difficulty (3) could be lifted if all functions with "unspecified" evaluation order actually picked a random evaluation order at runtime, letting user realize the arbitrariness of it by testing.)

I believe that "unspecified" in this context is mostly a cop-out from the designers to be dismissive when the users report bugs caused by bad library design. This is not the case in all circumstances: when you write an advanced API for advanced purposes, you can assume a certain level of discipline among its users. (I was going to write "it is fine if `Obj` carefully documents what is and isn't specified", oh well...). All your students are going to use `List.init`.

dra27 commented:
@gasche OK! What's your opinion on the merits of two functions, therefore, proposed earlier? If it does go that way, I might suggest `init_left` and `init_right` rather than `init` and `init_rev`?

gasche commented:
We use `rev_map`, so maybe `rev_init` would be better than `init_rev`. Nothing in particular against `{left,right}_init`, but won't users be looking for a function just named `init`? (I think someone could make the case that, in the interest of minimality, maybe we only should have one `init` function. I am not convinced that this is really an argument.)

dra27 commented:
Except that rev_map actually produces a different kind of list - the output of `rev_init` and `init` is always `[f 0; f 1; ... ]`. (it's also funny that `rev_init` constructs the list in order and `init` is the one which has to call `rev`!!)

Given your (acceptable!) fervour for least-surprise, perhaps *not* having init is good - someone using Merlin will see that there is init_left and init_right and maybe they'll do the thinking they should be?

gasche commented:
Apologies for being unclear. The `rev_init` I had in mind would return `[f (n - 1); ...; f 0]`, so that `rev (rev_foo f n) = foo f n` holds for both `map` and `init`. That said, I don't have much opinions on functions with a different name (those may be used by more informed users, especially if they only find them by reading the documentation).

dra27 commented:
Ah, I see! Any further thoughts @mshinwell and @nojb on this typically long stdlib discussion? :smile:

nojb commented:
Don't have much to add to the tail-recursive/reverse-order question, but personally I think having two functions is overkill for such a trivial function (and one that is not even used that much AFAIK).

bobzhang commented:
how about just the plain old one:

```ocaml
let init n f = Array.init n f |> Array.to_list
```
or inline `Array.init` manually.

For the evaluation order, it would cause a big surprise that the order is not from 0 to n

Richard-Degenne commented:
> this typically long stdlib discussion

Actually, this is very interesting, and since this is the first time I've been a part of it, I'm really glad to see such "discussions"  :smiley:

This is occuring to me just now, but the cases where `f` has side-effects where execution order is a problem might be relatively rare.

bluddy commented:
How about this option: make an List.init function that takes an optional ~rev boolean. The standard implementation will initialize each element in-order and do a final List.rev. This way, neophytes will find what they're looking for and it'll meet their expectations. The ~rev option will use an alternate path that initializes in reverse, and does no List.rev.

Richard-Degenne commented:
I'm not convinced with the name `rev`: I feel like it implies the function will return `[f (n-1); f (n-2); ...; f 0]`.

Drup commented:
My 2cents: in 99% of the use cases of List.init, it doesn't matter one way or the other because it will be on pure function with reasonably small lists. Pick either the tail-rec version or the left-to-right version, forget the `rec` non-sense, users who have specific needs can write the 3 line recursion function themselves.

Anyway, if you need lists big enough that you need tail-rec List.init, you should most probably be using generators or arrays.

One point in favor of left-to-right is that when/if we get tail rec modulo cons, we can turn it tail-rec without breaking compat.

bobzhang commented:
but that 1% will bite u in a hard way

xandkar commented:
What @bluddy said, but ```~order: [`Increasing | `Decreasing]```.

mookid commented:
@bluddy @ibnfirnas  the rest of the stdlib does not really make use of optional/named arguments. I agree strongly with @Drup : you can't count on a reasonable default for the evaluation order of arguments in OCaml in general - some will call that a feature. 

In my opinion we should implement one easy to read version (such as the current `List.map` implementation). Users with particular needs will need to plug their own specialized version when dealing with either overflow or side effects issues.

bobzhang commented:
For a stdlib, the default  should be minimal surprise not bare metal perf.  
If people do need bare metal perf, they should roll their own library, that is what I did for bucklescript and probably you should do if are very sensitive about perf .
That does not mean it would be slow, it should be reasonably fast, but when they are at odds with least surprise, we should definitely go with a more reasonable API. For example, this would be a **WAT for OCaml**
```ocaml
List.init 3 (fun i -> print_int i ; i)
```
```
2
1
0
```
and beginners see reverse order...
This is different from parameter evaluation order, it is well know that people should not rely on when reading`f a b`, the evaluation order of `a` or `b`, here it is totally different

lefessan commented:
Non-tailrec `List.map` was a huge mistake, and a non-tailrec `List.init` would probably be another one. Also, users are expecting the 0..(n-1) evaluation order, so we should keep that.

 Since the `List` module already depends on `Array`, the best trade-off with these constraints seems to be the standard `let init n f = Array.to_list (Array.init n f)`, since the list is only allocated once, and the array is usually much smaller than the space allocated by the reversed list.

Richard-Degenne commented:
Using `Array.init` in `list.ml` introduces a new dependency, though. Am I supposed to add something to the Makefile?

nojb commented:
You would need to change the order they are listed in the `Makefile`, since `list.cmo` currently comes before `array.cmo`.  However, I think it is better to simply inline the necessary function definitions.

mookid commented:
@lefessan if the design of `List.map` is such an issue, why isn't it fixed?

lefessan commented:
@mookid I think it's a matter of performance, devels are afraid of the cost of a tailrec version of `List.map`, for a function that is not supposed to be applied on huge lists anyway. The problem is that, often, you work with little lists all the time, until the day where the list is huge and breaks your application in production...

bluddy commented:
@lefessan @mookid I find Core's approach to be best in List.map -- loop non-tailrec with a bound, and once you reach that bound, continue in a tail-rec fashion. You get the best of both worlds. They also unroll the loops, but with Flambda that should be unnecessary.

Since List.init has access to the desired size of the list, it can threshold it even more easily, choosing either tail-recursion or no tail-recursion based on whether the size approaches stack capacity.

damiendoligez commented:
* About evaluation order: when I was young I was a big fan of unspecified evaluation order. Now I know it's a big mistake: it lays traps for the programmer for which the type system cannot help.
* About tail recursion: the function should just work, whatever the size of the list.
* About using `Array`: I don't like it, if only because it would limit the size to 2^24 on 32-bit machines (see previous point).
* About the complexity of the implementation: it's OK to have a relatively complex design. That's what the standard library is for: provide good functions to users so they don't have to roll their own almost-good versions.

So this is what I'd like: size threshold + tail-rec and non-tail-rec versions of the code.

damiendoligez commented on stdlib/list.ml:
```diff
@@ -56,6 +56,15 @@ let rec rev_append l1 l2 =
+  init_aux ((f n)::acc) (n-1) f
+
+let init len f =
+  if len = 0 then [] else
```
This line is not needed.

mookid commented:
@damiendoligez 

Something like

```ocaml
let rec rev_init_tail_rec acc i n f =
  if i = n then acc
  else let fi = f i in rev_init_tail_rec (fi::acc) (i + 1) n f

let rev_init args = rev_init_tail_rec [] 0 args
let rec init_loop i n f =
  if i = n then []
  else
    let fi = f i in
    if i < 10_000 then fi::init_loop (i + 1) n f
    else fi::(List.rev (rev_init_tail_rec [] (i + 1) n f))

let init args = init_loop 0 args
```

?

alainfrisch commented:
@damiendoligez I fully agree with what you said.  Would you support a proposal to reimplement e.g. `List.map` more efficiently and stack-overflow-proof in the stdlib?  One tricky part is that optimizing for the current backend (e.g. by manually unrolling the loop, or adding logic to switch to a tail-rec version after some iterations) could prevent inlining e.g. with flambda.

Richard-Degenne commented:
Ok, so I feel like we've reached a consensus here. I'll be pushing soon.

mookid commented on stdlib/list.ml:
```diff
@@ -56,14 +56,18 @@ let rec rev_append l1 l2 =
+
+let rec init_aux i n f =
+  if i >= n then []
+  else (f i)::init_aux (i+1) n f
```
Without explicit sequencing (`let fi = f i in...` in my message) this implementation fails to maintain the evaluation order spec.

In the top level:
```ocaml
# let _ = init_aux 0 5 print_int;;
43210- : unit list = [(); (); (); (); ()]
```

Richard-Degenne replied:
Agreed. I should take some time to make unit tests able to detect that.

mookid replied:
You can test the behavior before and after the threshold (currently 10_000) using a function such as 

```ocaml

fun i -> if i < 10 || i > 10_000 && i < 10_010 then print_endline (string_of_int i)

```


It would allow to cover most branches without committing a giant test output file.

gasche replied:
`i > 9_995` would be even better to test the threshold.

alainfrisch commented:
> About using Array: I don't like it, if only because it would limit the size to 2^24 on 32-bit machines (see previous point).

I'm not claiming it's necessarily a great idea to do it, but one could have a third implementation only used when the size is too large for a non-tail recursion, but small enough to fit in an array.  Or even create a list of chunks (large arrays), and then build the final list for it.  Assuming the list elements are immediate values, this would reduce by a factor of 3 the extra required temporary memory, and reduce the GC pressure significantly (but the write barrier would have a cost, too).

alainfrisch commented on stdlib/list.ml:
```diff
@@ -56,6 +56,21 @@ let rec rev_append l1 l2 =
 
+let rec init_tailrec_aux acc i n f =
+  if i >= n then acc
+  else init_tailrec_aux ((f i)::acc) (i+1) (n) f
```
You should remove the parenthesis around `n` (and also possibly around `f i`).

Richard-Degenne replied:
Agreed.

Commit:
Changed `List.init` behaviour

After discussion on [the feature's pull
request](https://github.com/ocaml/ocaml/pull/1034), `List.init` have
been updated.

- It now evaluates `f` from `0` to `n-1`;
- It is now non-tail-recursive if `n<=10_000`.
- It remains tail-recursive if `n>10_000`.

**Note** : The tail-recursiveness threshold is arbitrary and may be up
for discussion.

dra27 commented on stdlib/list.ml:
```diff
@@ -56,6 +56,21 @@ let rec rev_append l1 l2 =
 
+let rec init_tailrec_aux acc i n f =
+  if i >= n then acc
+  else init_tailrec_aux (f i::acc) (i+1) n f
```
Spaces around the `::` (see other instances in the file)

dra27 commented on stdlib/list.ml:
```diff
@@ -56,6 +56,21 @@ let rec rev_append l1 l2 =
+  if i >= n then []
+  else
+    let fi = f i in
+    fi::init_aux (i+1) n f
```
Again, for consistency elsewhere could variable `fi` be `r`, please (and spaces around the `::`)

dra27 commented on stdlib/list.mli:
```diff
@@ -73,6 +73,11 @@ val nth_opt: 'a list -> int -> 'a option
 (** List reversal. *)
 
+val init : int -> (int -> 'a) -> 'a list
+(** [List.init len f] is [f 0; f 1; ...; f (n-1)].
```
This comment is now longer true - the clearest is probably to add ` (evaluated left to right)`

The `@since` and `@raise` changes still need making.

gasche replied:
While we are at it, you use `len` on the left and `n` on the right.

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -73,6 +73,11 @@ val nth_opt: 'a list -> int -> 'a option
+(** [List.init len f] is [f 0; f 1; ...; f (n-1)].
+
+   Raise [Invalid_argument] if [n < 0].*)
+
```
Will need same changes as in list.mli

dra27 reviewed (changes requested):
You merged a change in from trunk - would you be able to rebase to get rid of that merge commit? Other than that, just a little tidying up here and there.

It would be good to add an extra test for the threshold as @mookid and @gasche suggested.

Commit:
Fixed execution order

Commit:
Improved formatting

Commit:
Improved formatting and documentation

mshinwell commented:
@damiendoligez Your comments didn't seem to address the question of "forwards" versus "reverse" evaluation order here.

bluddy commented:
@mshinwell: "About evaluation order: when I was young I was a big fan of unspecified evaluation order. Now I know it's a big mistake: it lays traps for the programmer for which the type system cannot help."

damiendoligez commented:
@mshinwell Left-to-right is less surprising, so I think it's what we should specify.

damiendoligez commented on stdlib/list.ml:
```diff
@@ -56,6 +56,21 @@ let rec rev_append l1 l2 =
+
+let init len f =
+  if len < 0 then invalid_arg "List.init" else
+  if len > 10_000 then rev (init_tailrec_aux [] 0 len f)
```
Is 10000 small enough, even with the tiny default stack provided by Windows?

Richard-Degenne replied:
I used the arbitrary value suggested by @mookid, but, as stated in my commit message, it is up for discussion.

dra27 replied:
@damiendoligez - the default stack may be tiny, but doesn't flexlink always override it? I certainly have no recursion problems with lists up to 100000 on the couple of Windows ports I had immediately to hand?

Commit:
Added evaluation order unit tests

Richard-Degenne commented:
Any input on the current state of the pull request?

dra27 commented:
@Richard-Degenne There is still, I'm afraid, another tiny tweak to the .mli comments! I'm not sure if it matters, but in an `@raise` the exception name should not be wrapped in brackets - so it should be `@raise Invalid_argument if [len < 0]` - ocamldoc requires the first word after `@raise` to be the exception name and formats it accordingly.

It's also important to get an answer from the CI - please could you rebase to remove the clashing merge commit?

dra27 commented on testsuite/tests/lib-list/test.ml:
```diff
@@ -0,0 +1,56 @@
+      if x = 9999 then result := false
+      else if x = 10000 then result := true
+    )
+  in assert !result
```
This doesn't adequately test all code-paths, an improved version:

```ocaml
let () =
  let test n =
    let result = ref false in
    let _ = List.init n (fun x -> result := (x = n - 1)) in
    assert !result
  in
    (* Test cases must be either side of the threshold in stdlib/list.ml *)
    test 10000;
    test 10001
;;
```

It also be sensible to include a comment in list.ml that if the 10_000 is changed that the test needs updating.

Commit:
Tweaked documentation formatting

Commit:
Changed unit tests

alainfrisch commented:
Squashed and merged in trunk. Thanks to @Richard-Degenne  and to everyone involved in the discussion!

murmour commented on testsuite/tests/lib-list/test.ml:
```diff
@@ -0,0 +1,47 @@
+  assert (not (List.exists (fun a -> a > 9) l));
+  assert (List.exists (fun _ -> true) l);
+;;
+
```
Um. Shouldn't the following suffice? We aren't testing `List.exists` here.
```ocaml
assert ((List.init 10 (fun x -> x)) = [ 0; 1; 2; 3; 4; 5; 6; 7; 8; 9 ])
```

dra27 replied:
FWIW, yes we are!

infinity0 commented:
Just wanted to document the fact that this PR breaks pre-4.06.0 code like:

```ocaml
let f init x y z =
  let open List in
  (* do something with `init` *)
```

You'll get an error message like

```
Error: This expression has type int -> (int -> 'weak7) -> 'weak7 list
       but an expression was expected of type (your actual type) = (your actual type)
```

Maybe this will help people googling for the error message.

infinity0 commented:
(Ofc the fix is easy, just rename your `init` to something else.)

dra27 commented:
Sorry you got stung, though this shows the general weakness (in my not remotely humble opinion) of opening stdlib modules! `List.( ... )` will probably be safer if you use several symbols from `List` in a single expression and otherwise for single symbols, `List.length` is much clearer than `length`, etc.

hhugo commented on stdlib/list.ml:
```diff
@@ -56,6 +56,21 @@ let rec rev_append l1 l2 =
+
+let init len f =
+  if len < 0 then invalid_arg "List.init" else
+  if len > 10_000 then rev (init_tailrec_aux [] 0 len f)
```
This will stack-overflow when compiled to JavaScript with js_of_ocaml (because the javascript stack is much smaller). Any chance to lower the limit to ~100 ?

Gbury replied:
Maybe the constant could depend on `Sys.backend_type` (cf https://github.com/ocaml/ocaml/blob/trunk/stdlib/sys.mli#L99 ) ?

dra27 replied:
It's not wholly clear from its status, but this PR was merged, so please could this go in a new PR or in a Mantis report?

damiendoligez replied:
Reported at [MPR#7764](https://caml.inria.fr/mantis/view.php?id=7764).

hhugo replied:
https://github.com/ocaml/ocaml/pull/1697
